Create EIP-4361: Sign-In with Ethereum (#4361)
ethereum/EIPs: EIPS/eip-4361.md

> ---

Reviewer: 0x14D6485F5E3B109230563b1d443fD6B16539955c

---
eip: 4361
title: Sign-In with Ethereum. 
description: Standardizes off-chain authentication for Ethereum accounts to establish sessions.
author: Wayne Chang (@wyc), Brantly Millegan (@brantlymillegan), Nick Johnson (@Arachnid)

> resolution: <TBD>

Reviewer 2: ```suggestion
```
This is not an EIP frontmatter field.

Author: @MicahZoltu from https://eips.ethereum.org/EIPS/eip-1#eip-header-preamble
```
` * resolution:` a url pointing to the resolution of this EIP
```

Is there a more updated version you could point me to that reflects its omission? Thanks

Reviewer 2: https://raw.githubusercontent.com/ethereum/EIPs/master/eip-template.md I'll get a PR out to fix EIP-1 to match.

Ignore NizamShah7, it is just a spam account.

Author: Removed in https://github.com/ethereum/EIPs/pull/4361/commits/0975d6c0d5a610dd8209d9911c715702c8600c62 , thanks for your advice! Feel free to reopen if there's anything else.

Already, many services support workflows to authenticate Ethereum accounts using message signing, such as to establish a cookie-based web session which can manage privileged metadata about the authenticating address. This is an opportunity to standardize the sign-in workflow and improve interoperability across existing services, while also providing wallet vendors a reliable method to identify signing requests as Sign-In with Ethereum requests for improved UX.

This work is sponsored by the Ethereum Foundation and Ethereum Name Service (ENS). It is being developed in the open through a series of recorded community calls and public repositories, and its development is informed by over twenty user interviews with a focus on currently-in-production uses, related prior EIPs, and fits within product roadmaps.

## Specification
Sign-In with Ethereum works as follows:

> 1. The wallet presents the user with a structured plaintext message or equivalent interface for signing with the EIP-191 signature scheme (string prefixed with `\x19Ethereum Signed Message:\n<length of message>`). The `message` MUST incorporate an Ethereum `address`, `version` of the message, `uri` for scoping, `nonce` acceptable to the server, and `issued-at` timestamp.

Reviewer 3: I tend to agree with @PhABC here. It's not completely obvious the requirement for a `nonce` and what benefit nonce actually providers for "the most developers" category.

Sure, for med/high security situations, it's important to re-authenticate a user before allowing roles based access control to resources (compute or otherwise), but I would argue "generic" authentication uses a timestamps for session management and if a re-authentication is required you invalidate the active session signature/JWT

Could easily see most services constantly using the `0` or `1` nonce, because they have too, but don't have a real need for it.

If a server is going to be "stateful it would be good to research what's the most common techniques for invalidation of active sessions. To my knowledge a lot services will just invalidate a previous session cookie and require a user re-authenticate, which nullifies the requirement for a nonce anyways, because they can be handled via timestamp parameters.

> 1. The wallet presents the user with a structured plaintext message or equivalent interface for signing with the EIP-191 signature scheme (string prefixed with `\x19Ethereum Signed Message:\n<length of message>`). The `message` MUST incorporate an Ethereum `address`, `version` of the message, `uri` for scoping, `nonce` acceptable to the server, and `issued-at` timestamp.

Reviewer 4: Hmm maybe, I'm missing something here but can someone explain why login is necessary in a stateless app interaction? I want to make sure I understand the architecture of how this is being used since I think I may be missing the point in why this would be useful.

> 1. The wallet presents the user with a structured plaintext message or equivalent interface for signing with the EIP-191 signature scheme (string prefixed with `\x19Ethereum Signed Message:\n<length of message>`). The `message` MUST incorporate an Ethereum `address`, `version` of the message, `uri` for scoping, `nonce` acceptable to the server, and `issued-at` timestamp.

Reviewer 3: @kdenhartog I think by stateless what's being expressed is no database in the authentication layer. Traditionally with a signed JWT you can get away with authenticating signatures without pinging a database to validate a session. If the signature is valid and the timestamp hasn't expired it's good to go.

> 1. The wallet presents the user with a structured plaintext message or equivalent interface for signing with the EIP-191 signature scheme (string prefixed with `\x19Ethereum Signed Message:\n<length of message>`). The `message` MUST incorporate an Ethereum `address`, `version` of the message, `uri` for scoping, `nonce` acceptable to the server, and `issued-at` timestamp.

Reviewer 4: >Traditionally with a signed JWT you can get away with authenticating signatures without pinging a database to validate a session. If the signature is valid and the timestamp hasn't expired it's good to go.

Yes, but that's because the resource server has an inherent binding to the authorization server because it's directly retrieving the token from the OP in the authorization code flow. So in other words the application itself can be managed in a stateless manner because the OP can be trusted to not maliciously replay the token due to the trust between the two servers in that architecture. This means the nonce is unnecessary to be used for replay protection and it's only being used for the second purpose which is to associate a Client session with an ID Token. However, because the authorization code is also being used to associate the client session with the ID Token, the nonce no longer is necessary in this flow.

The issue here is that the resource server can't trust the authorization server in this case and so the architecture of SIWE is more in line with the implicit code flow which does require the usage of a nonce to be sent in the request. This is in section [3.2.2.1 of the OIDC spec](https://openid.net/specs/openid-connect-core-1_0.html#ImplicitAuthRequest).

Building on this the spec says in section [3.2.2.11 of the OIDC Core spec](https://openid.net/specs/openid-connect-core-1_0.html#ImplicitIDTValidation) it states:

> if a nonce value was sent in the Authentication Request, a nonce Claim MUST be present and its value checked to verify that it is the same value as the one that was sent in the Authentication Request. The Client SHOULD check the nonce value for replay attacks. The precise method for detecting replay attacks is Client specific. 

So in other words, because there's not an inherent trust between the RP and the wallet a client MUST be used. If an app wanted to go stateless with the current design they would most likely need to utilize the OAuth server that Spruce also has committed to so that the RP could establish a trust and defer authn and authz events to the OP.

> 1. The wallet presents the user with a structured plaintext message or equivalent interface for signing with the EIP-191 signature scheme (string prefixed with `\x19Ethereum Signed Message:\n<length of message>`). The `message` MUST incorporate an Ethereum `address`, `version` of the message, `uri` for scoping, `nonce` acceptable to the server, and `issued-at` timestamp.

Reviewer 3: Sorry... I'm a little confused. What resource server? At the end of the day of this is minimally self-signed JWT-like token that's rooted to external sources via a verifiable address.

What are we getting from adopting that OIDC spec when we're removing the IDP and RP from the equation?

> 1. The wallet presents the user with a structured plaintext message or equivalent interface for signing with the EIP-191 signature scheme (string prefixed with `\x19Ethereum Signed Message:\n<length of message>`). The `message` MUST incorporate an Ethereum `address`, `version` of the message, `uri` for scoping, `nonce` acceptable to the server, and `issued-at` timestamp.

Reviewer 3: Hmmm... yeah I just don't see why that matters with the self-signing of the authentication token. The trust is from signature verification and `issuedAt` key, which essentially acts as `nonce` so why add a field to semi-adopt the ODIC spec?

> 1. The wallet presents the user with a structured plaintext message or equivalent interface for signing with the EIP-191 signature scheme (string prefixed with `\x19Ethereum Signed Message:\n<length of message>`). The `message` MUST incorporate an Ethereum `address`, `version` of the message, `uri` for scoping, `nonce` acceptable to the server, and `issued-at` timestamp.

Reviewer 4: It's not about trying to adopt the OIDC spec. I only used that as an example in order to showcase this is standard for protocol designs like this and to highlight that trust boundaries between the different software systems matter here. In this particular case due to the nature in which the trust works, the appropriate method to be used is to require the nonce in order to allow the protocol to be used in a broader set of use cases without requiring dApp developers to have to assess their threat model each time and decide if it's necessary in order to support 1 architecture which I'm still not convinced makes legitimate sense.

> 1. The wallet presents the user with a structured plaintext message or equivalent interface for signing with the EIP-191 signature scheme (string prefixed with `\x19Ethereum Signed Message:\n<length of message>`). The `message` MUST incorporate an Ethereum `address`, `version` of the message, `uri` for scoping, `nonce` acceptable to the server, and `issued-at` timestamp.

Author: Thanks for your comments, if you continue to feel strongly about this please continue the discussion in ETH Magicians.

```

### Informal Message Template
A Bash-like informal template of the full message is presented below for readability and ease of understanding. Field descriptions are provided in the following section. A full ABNF description is provided in the section thereafter.
```
${domain} wants you to sign in with your Ethereum account:
${address}

${statement}

URI: ${uri}
Version: ${version}
Chain ID: ${chain-id}
Nonce: ${nonce}
Issued At: ${issued-at}
Expiration Time: ${expiration-time}
Not Before: ${not-before}
Request ID: ${request-id}
Resources:
- ${resources[0]}
- ${resources[1]}
...
- ${resources[n]}
```
### Message Field Descriptions
- `domain` is the RFC 4501 `dnsauthority` that is requesting the signing.
- `address` is the Ethereum address performing the signing conformant to capitalization encoded checksum specified in EIP-55 where applicable.
- `statement` (optional) is a human-readable ASCII assertion that the user will sign, and it must not contain `'\n'`.
- `uri` is an RFC 3986 URI referring to the resource that is the subject of the signing (as in the _subject of a claim_).
- `version` is the current version of the `message`, which MUST be `1` for this specification.

> - `chain-id` is the EIP-155 Chain ID to which the session is bound, and the network where Contract Accounts must be resolved.

Reviewer 3: Why is the `chain-id` not an optional parameter? Also why is the session being bound to chainId to start with?

What does `resolved` in thus context mean for Contract Accounts?

Author: Yes, it would be required for EIP-1271 based signing. Furthermore, by principle of least privilege, we wanted to strongly push implementers to lock the session to a specific `chain-id` or think carefully about its interpretation otherwise. I think keeping a required `chain-id` reduces one conditional check (for smart contract wallet or EOA) and also achieves the latter goal. Does that make sense?

Reviewer 4: The flip side of this is that it requires me to re-authenticate each time I change networks in my wallet with the current UX flows. E.g. say I want to use sushiswap on mainnet and then decide to use a L2 solution I'd have to re-authenticate after changing networks which doesn't make sense to me. Wouldn't this lead to greater consent fatigue which is already pretty high as is? Furthermore, is it likely that we'd have the same account on different networks owned by different people?

Author: @kdenhartog An implementer is free to interpret a `chain-id` of `0` as across all networks for example, but that's their decision to make. From today's application, it does appear that networks aiming for EVM-compat indeed re-use the same account. The privacy and security implications of this are worth noting but out of scope for this spec.

Reviewer 4: It sounds like it's being made in scope though because the chain-id is being required so that's a bit confusing to me. Requiring it's usage, but not stating why it's necessary nor placing additional normative requirements on it seems a bit odd to me and is opening the door for potential interoperability issues based on wallets and dApps making different assumptions.

Author: Would you onboard with adding the recommendation of `chain-id` to be `0` if it's a "don't care"?

Reviewer 4: Yeah I think normatively defining that would be a good way to prevent interop concerns. Strong +1 from me. Would chain-id 0 also need to be registered with EIP 115?

Author: Hm, I don’t think it’s normatively defined in EIP-155, could be better to add the character `*` as a valid atom for this to avoid any future woes, okay with this?

Author: It also begs the question of smart contract wallets being able to sign cross chain…in earlier iterations we had a list of `chain-id` but found that to be overkill…

Reviewer 4: I'm good with a `*` character, but I think it may encounter some weird logic in the way EIP-115 was originally intended, but I'm not certain there's overlap at this point. Particularly this clause in EIP-115 makes me think it could break some logic because it's not a number type:

> If block.number >= FORK_BLKNUM and v = CHAIN_ID * 2 + 35 or v = CHAIN_ID * 2 + 36

>It also begs the question of smart contract wallets being able to sign cross chain…in earlier iterations we had a list of chain-id but found that to be overkill…

Yeah that seems like it would be a really useful feature, but I'd assume based on your initial response that including this may be more scope creep than you'll want to take on at this point. I'm good with leaving the cross contract signing off for now in order to simplify, but there's definitely tradeoffs that will come with this in terms of limited capabilities here.

Reviewer 3: > Would you onboard with adding the recommendation of `chain-id` to be `0` if it's a "don't care"?

Doesn't that imply `chain-id` is optional and the presumed default is 0? So no need to redeclare `0` with `0`.

Author: @kamescg I propose we keep it due to the reasoning here: https://github.com/ethereum/EIPs/pull/4361#discussion_r755524905

We can continue this discussion in the ETH Magicians forum if you feel strongly about it.

- `nonce` is a randomized token used to prevent replay attacks, at least 8 alphanumeric characters.
- `issued-at` is the ISO 8601 datetime string of the current time.
- `expiration-time` (optional) is the ISO 8601 datetime string that, if present, indicates when the signed authentication message is no longer valid.
- `not-before` (optional) is the ISO 8601 datetime string that, if present, indicates when the signed authentication message will become valid.
- `request-id` (optional) is an system-specific identifier that may be used to uniquely refer to the sign-in request.
- `resources` (optional) is a list of information or references to information the user wishes to have resolved as part of authentication by the relying party. They are expressed as RFC 3986 URIs separated by `"\n- "`.

### ABNF Message Format

> The `message` to be signed MUST conform with the following Augmented Backus–Naur Form (ABNF, RFC 5234) expression (note that `%s` denotes case sensitivity for a string term, as per RFC 7405).

Reviewer 5: Why is it preferable to have a new invented language and not some binary format, which is easier to parse (CBOR or even ERC-712)?

Is the only reason because the signed message will be "readily" displayed in wallets without implementing support for it? If so, I think hardware wallets may be a big exception to that where some other more easily parsable format is more likely to gain support.

Author: Hi @axic, thanks for your concern. That is indeed the main reason, so that applications may adopt this specification without full buy-in from wallet vendors or significant degradation of their user experiences today. In our proposal response to the RFP, we actually specified EIP-712 as the signing format, but found the user experience across many wallets to be much worse than using EIP-191. Formats like CBOR, EIP-712, protobuf, etc. indeed specify structuring for data, but hardware wallets have spotty support for presentation of EIP-712 requests to the users already (please see ongoing issues https://github.com/LedgerHQ/app-ethereum/issues/105 and https://github.com/trezor/trezor-firmware/pull/1568). We believe that if adoption of this specification relies on all wallet vendors upgrading how they do signing first, it is far less likely to see success. While pure technical merit is important to consider, adoption depends far more on talking to downstream users and understanding their concerns, what they are likely to truly adopt, etc. We have completed over 30 interviews towards this conclusion. We welcome anyone to let us know of any further user research in support or incongruent with this claim. Hope that helps!

Author: Hi, just wanted to follow up on this thread. I understand that people like structured signing and this decision may be unpopular, but we did the work and this is how it looks like in wallets today: https://blog.spruceid.com/sign-in-with-ethereum-wallet-support-eip-191-vs-eip-712/

Applications we spoke to are not happy to degrade their UX significantly, and people have little idea what they are signing if confronted with a JSON blob. Therefore, I think EIP-191 is still the right move here. 

If you'd like to consider CBOR, EIP-712, protobuf, etc., I would need some evidence that indeed broad wallet support will come for these soon so users can see what they're signing. We added a `version` term to allow this as an upgrade path in the future, but for this first iteration of the spec it seems premature. Thanks so much for your review and consideration.

Already, many services support workflows to authenticate Ethereum accounts using message signing, such as to establish a cookie-based web session which can manage privileged metadata about the authenticating address. This is an opportunity to standardize the sign-in workflow and improve interoperability across existing services, while also providing wallet vendors a reliable method to identify signing requests as Sign-In with Ethereum requests for improved UX.

This work is sponsored by the Ethereum Foundation and Ethereum Name Service (ENS). It is being developed in the open through a series of recorded community calls and public repositories, and its development is informed by over twenty user interviews with a focus on currently-in-production uses, related prior EIPs, and fits within product roadmaps.

## Specification
Sign-In with Ethereum works as follows:

> 1. The wallet presents the user with a structured plaintext message or equivalent interface for signing with the [EIP-191](https://eips.ethereum.org/EIPS/eip-191) signature scheme (string prefixed with `\x19Ethereum Signed Message:\n<length of message>`). The `message` MUST incorporate an Ethereum `address`, `version` of the message, `uri` for scoping, `nonce` acceptable to the server, and `issued-at` timestamp.

Reviewer 6: Does reusing `\x19Ethereum Signed Message:\n<length of message>` pose a security risk?

If an attacker on a visually similar phishing domain uses the existing `eth_signMessage` endpoint to sign a message, do we expect existing wallets to display a warning?

Reviewer 7: I don't quite understand the question here related to the prefix (as mentioned in EIP-191) and domains?

If we want the wallet to display a warning then we'd need to require the wallet to parse a domain from the message and check this against the domain of the website in use. Should wallets do that? Sure, it would be nice and good UX as far as I'm concerned.

But how does this relate to the prefix?

Reviewer 6: If you use a separate prefix, existing RPC endpoints like `eth_signMessage` will generally refuse to sign, so the wallets don't need to implement any special checks on _all_ signed messages, only on ones that use the new prefix.

If you use the same prefix, wallets have to scan all `eth_signMessage` bodies for messages that parse as EIP-4361 compatible, and then check the domain.

Author: @SamWilsn Thanks for your comments! We should encourage wallet vendors to display a separate interface once they have identified a valid SIWE message, but it's true they'd have to interpret the whole thing in the current form. Two possible improvements here, not mutually exclusive:

1. We could alternatively require the message to begin with `"EIP-4361. "` to allow quick prefix determination for parsers prior to attempting to parse any of the grammar, but I think the message should be short enough as not to pose a big risk. This would potentially set a precedence on how future signing requests will differentiate themselves, inadvertently setting a precedent for future EIP-191 extensions. It could also slightly degrade the user experience, but arguably also improve it by giving users a term that is likely to turn up with useful results in a search engine.

2. We can add a SHOULD clause to the spec that cannot be easily tested like: "Wallet vendors SHOULD display a confirmatory indicator if a valid EIP-4361 message is detected, and wallets MAY take reasonable steps to proactively identify similar but non-conformant messages to alert the user accordingly."

Thoughts on these?

Author: Also, note on the separate prefix--I don't think most wallets would be able to display the message cohesively if we removed the signing prefix `\x19Ethereum Signed Message:\n<length of message>` for wallets that don't make changes explicitly to support this spec. The goal is that applications should be able to create conforming implementations without forcing wallet vendors to do anything without substantially degrading the user experience. This is an important consideration for adoption, and at the same time your observation is valid.

Reviewer 6: I'm not a fan of (1) since it adds extra gibberish to the message that will make users even more likely to just blindly sign.

Is the `wants you to sign in with your Ethereum account` portion fixed and required? If so, that's probably sufficient for wallets to scan for.

I do think (2) is a good security recommendation. Maybe something as simple as: "Wallet software SHOULD display a high friction confirmatory indicator if the string `wants you to sign in with your Ethereum account` appears anywhere in an EIP-191 message signing request unless the message is a valid according to EIP-4361."

Author: Thanks! I will make these changes, okay to add you as a contributor for your useful comments throughout?

Reviewer 6: Sure, that's fine!

Reviewer 3: I also found the `wants you to sign in with your Ethereum account` to be an "interesting" part of the specification.

How does this work with multiple languages in regards to the specification? Ideally the specification has no dependency on the English language and can be used in a universal setting. Can this be replaced with translated equivalents, and if so will wallet vendors be expected to parse for those strings as well? 

This is kind of secondary, but I found it just a little awkward to parse the current message template. Here is just a quick, naive approach. 

https://gist.github.com/kamescg/b0f2875881ff15ec976f6976cedca885

Reviewer 3: > Maybe something as simple as: "Wallet software SHOULD display a high friction confirmatory indicator if the string wants you to sign in with your Ethereum account appears anywhere in an EIP-191 message signing request unless the message is a valid according to EIP-4361."

The best solution definitely feels like parsing the `eth_signMessage` to detect message `types` is the most flexible long-term solution. Also, it's hard to believe this will be the last EIP to use string parsing to detect the message `intention`.

Because of this I feel like detecting such a "generic message" like `wants you to sign in with your Ethereum account` is not the most scalable approach to dynamically detecting all incoming `eth_signMessage` potential future types.

In my opinion it would probably be best to drop the `wants you to sign in with your Ethereum account` because that's implied in the singing 😉  and replace it with a `type` and `language` keys that can be parsed by incoming wallets and also be used by future standards who need to have a `StructuredDataIsh` interface while still using the `eth_signMessage` call.

Easy to parse and identify.

And the intention message maybe has to be replaced with `Domain` and `Address` keys?

Also just curious as to why the `statement` param isn't a key/value pair like everything else? Is a bit of an odd data structure when the message has to be dynamically parsed and found by searching for the two surrounding `\n` on empty lines.

**Proposed Changes**
```
type: EIP-4361
language: en
Domain: http://localhost:3000 
Address: 0x000000000000000000000000000000000000dEaD
Statement: http://opensea.io/tos
URI: http://localhost:3000/authenticate
Version: 1
Chain ID: 1
Nonce: 12345678
Issued At: 2021-11-20T10:10:10Z
Not Before: 2021-11-19T10:10:10Z
Expiration Time: 2021-11-21T10:10:10Z
Request ID: 2
Resources:
 https://ipfs.com/hash
```

Author: Hi Kames, the example you present is indeed more normalized, but I am concerned about readability to end users. We gathered feedback from over 30+ teams in the space over the past several weeks, and they were okay with the proposed format. We welcome your independent research to see if they prefer the format you propose for their users over the current one if you wish to move it forward.

Additionally, I would consider solving for a "scalable approach to dynamically detecting all incoming eth_signMessage potential future types" is out of scope for this spec.

Thanks again for your comments, and I look forward to continuing the conversation in ETH Magicians if you continue to feel strongly about this.

### Example message to be signed
```
service.org wants you to sign in with your Ethereum account:
0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2

I accept the ServiceOrg Terms of Service: https://service.org/tos

URI: https://service.org/login
Version: 1
Nonce: 32891756
Issued At: 2021-09-30T16:25:24Z
Chain ID: 1
Resources:
- ipfs://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq/
- https://example.com/my-web2-claim.json
```

### Informal Message Template
A Bash-like informal template of the full message is presented below for readability and ease of understanding. Field descriptions are provided in the following section. A full ABNF description is provided in the section thereafter.
```
${domain} wants you to sign in with your Ethereum account:
${address}

${statement}

URI: ${uri}
Version: ${version}
Nonce: ${nonce}
Issued At: ${issued-at}
Expiration Time: ${expiration-time}
Not Before: ${not-before}
Request ID: ${request-id}
Chain ID: ${chain-id}
Resources:
- ${resources[0]}
- ${resources[1]}
...
- ${resources[n]}
```
### Message Field Descriptions

> - `domain` is the [RFC 4501](https://datatracker.ietf.org/doc/html/rfc4501) `dnsauthority` that is requesting the signing.

Reviewer 6: Should this be `dnsname`? Seems like `dnsauthority` is the server you're resolving against.

```suggestion
- `domain` is the [RFC 4501](https://datatracker.ietf.org/doc/html/rfc4501) `dnsname` that is requesting the signing.
```

> If an authority ("dnsauthority") is given in the URI string, this indicates the server that should receive the DNS query. Otherwise, the default DNS server should receive it.

_[source](https://datatracker.ietf.org/doc/html/rfc4501#page-4)_

Author: Agreed with your comment on naming to capture intents, especially as they are interpreted for TLS. However, we are setting up an application level session and not necessarily a comms channel. So, I think we should keep `dnsauthority`, which includes the host and port, to provide an opportunity for more granular binding. I'm thinking about the case where there may be multiple SIWE services exposed on the same host, but separated by port. This would allow the signing to be bound to an exact service and serve as an additional layer of protection on top of the nonce. Does that reasoning make sense?

Reviewer 6: I disagree with your use of:

> `domain` is the [RFC 4501](https://datatracker.ietf.org/doc/html/rfc4501) `dnsauthority` that is requesting the signing.

In RFC 4501, `dnsauthority` is _who_ you are contacting to look up a domain, and not a domain itself.

If your intent is to capture the "part of the URL" which represents the server side, I'd recommend switching to `authority` from [RFC 3986 (Section 3)](https://datatracker.ietf.org/doc/html/rfc3986#section-3) or `host` from [WhatWG's URL Specification](https://url.spec.whatwg.org/#hosts-(domains-and-ip-addresses)).

Author: Great suggestion, I think we'll go with one of those. Thanks for sharing your knowledge, I didn't think to look at WhatWG for a similar thing.

Author: Updated in https://github.com/ethereum/EIPs/pull/4361/commits/1415bfb31f5647b428fd06e9c0f74f78d589bebd with credits too.

- `address` is the Ethereum address performing the signing conformant to capitalization encoded checksum specified in [EIP-55](https://eips.ethereum.org/EIPS/eip-55) where applicable.

> - `statement` (optional) is a human-readable ASCII assertion that the user will sign, and it must not contain `'\n'`.

Reviewer 6: ```suggestion
- `statement` (optional) is a human-readable ASCII assertion that the user will sign, and it must not contain `'\n'` (the byte `0x0a`).
```

Author: This is great clarification, thank you!

> - `statement` (optional) is a human-readable ASCII assertion that the user will sign, and it must not contain `'\n'`.

Reviewer 6: I might suggest also excluding `:` (colon) to avoid confusing the assertion with the subsequent metadata.

Author: Great suggestion. I considered this, but some implementations including OpenSea may want to include a URL to their terms of service or similar, which would require `:` to construct. Is this exception reasonable enough to you to warrant its inclusion?

Reviewer 6: They can percent-encode it right?

Author: I will keep what's there for now, due to our user research. If you feel strongly about this, please continue the conversation in ETH Magicians. Thanks again.

- `uri` is an [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986) URI referring to the resource that is the subject of the signing (as in the _subject of a claim_).
- `version` is the current version of the `message`, which MUST be `1` for this specification.
- `nonce` is a randomized token used to prevent replay attacks, at least 8 alphanumeric characters.
- `issued-at` is the [ISO 8601](https://datatracker.ietf.org/doc/html/rfc3339) datetime string of the current time.

> - `expiration-time` (optional) is the [ISO 8601](https://datatracker.ietf.org/doc/html/rfc3339) datetime string that, if present, indicates when the signed authentication message is no longer valid.

Reviewer 6: Maybe rename to `not-after` for symmetry with `not-before`?

---

Does if the time is exactly equal to `expiration-time`, is the signature valid?

Reviewer 3: Great naming suggestion!

Also, in regards to is the "signature valid" I think that's up to the implementer of the authentication service, no? Can't really be enforced in the message schema.

Author: Thanks for your suggestion, I like it a lot semantically. However, I think I'd prefer to stick to `expiration-time` to keep similarity to JWTs (https://datatracker.ietf.org/doc/html/rfc7519) which would make it more ergonomic to developers already familiar with the terminology there. A lot of the spec was inspired by fields set as precedence in JWTs. Does that make sense?

Author: Interpreting your emoji as acceptance, feel free to reopen if you'd like to talk about this more. Thanks!

This work is sponsored by the Ethereum Foundation and Ethereum Name Service (ENS). It is being developed in the open through a series of recorded community calls and public repositories, and its development is informed by over twenty user interviews with a focus on currently-in-production uses, related prior EIPs, and fits within product roadmaps.

## Specification
Sign-In with Ethereum works as follows:

1. The wallet presents the user with a structured plaintext message or equivalent interface for signing with the [EIP-191](https://eips.ethereum.org/EIPS/eip-191) signature scheme (string prefixed with `\x19Ethereum Signed Message:\n<length of message>`). The `message` MUST incorporate an Ethereum `address`, `version` of the message, `uri` for scoping, `nonce` acceptable to the server, and `issued-at` timestamp.
2. The signature is then presented to the server, which checks the signature's validity and message content.

> 3. Additional fields, including `expiration-time`, `not-before`, `request-id`, `chain-id`, and `resources` may be incorporated as part of authentication for the session.

Reviewer 8: Could we consider the flexibility to support additional AuthN claims, for when it makes sense for the application? I can see some value in being able to switch between "wallet profiles" when it makes sense.

For example in a typical B2B App I'd also expect "department", or in another example, like a game, I could own multiple (in-game) characters under my wallet, would I be able to set an app-specific custom claim eg "character_name"?

Author: Hi epappas, please see the Resources section to allow this degree of flexibility. We just want to achieve ETH-based authn with this spec, not anything past that, though we are trying to keep that open ended for other implementers. Please feel free to continue the conversation in ETH Magicians if you feel strongly about it. Thanks.

## Motivation
When signing in to popular non-blockchain services today, users will typically use identity providers (IdPs) that are centralized entities with ultimate control over users' identifiers, for example, large internet companies and email providers. Incentives are often misaligned between these parties. Sign-In with Ethereum offers a new self-custodial option for users who wish to assume more control and responsibility over their own digital identity.

Already, many services support workflows to authenticate Ethereum accounts using message signing, such as to establish a cookie-based web session which can manage privileged metadata about the authenticating address. This is an opportunity to standardize the sign-in workflow and improve interoperability across existing services, while also providing wallet vendors a reliable method to identify signing requests as Sign-In with Ethereum requests for improved UX.

> This work is sponsored by the Ethereum Foundation and Ethereum Name Service (ENS). It is being developed in the open through a series of recorded community calls and public repositories, and its development is informed by over twenty user interviews with a focus on currently-in-production uses, related prior EIPs, and fits within product roadmaps.

Reviewer 9: Usually we don't include this kind of information in an EIP. I recommend removing it and instead posting it in the `discussions-to` thread.

Author: Addressed in https://github.com/ethereum/EIPs/pull/4361/commits/e59daacf95ecf9747b47921f7ce7b865eb139f75

## Specification
Sign-In with Ethereum works as follows:

> 1. The wallet presents the user with a structured plaintext message or equivalent interface for signing with the [EIP-191](https://eips.ethereum.org/EIPS/eip-191) signature scheme (string prefixed with `\x19Ethereum Signed Message:\n<length of message>`). The `message` MUST incorporate an Ethereum `address`, `version` of the message, `uri` for scoping, `nonce` acceptable to the server, and `issued-at` timestamp.

Reviewer 9: ```suggestion
1. The wallet presents the user with a structured plaintext message or equivalent interface for signing with the [EIP-191](./eip-191.md) signature scheme (string prefixed with `\x19Ethereum Signed Message:\n<length of message>`). The `message` MUST incorporate an Ethereum `address`, `version` of the message, `uri` for scoping, `nonce` acceptable to the server, and `issued-at` timestamp.
```
Please use relative links to other EIPs.

Reviewer 9: (there are other instances in the EIP that would need to be updated)

Author: Addressed in https://github.com/ethereum/EIPs/pull/4361/commits/e59daacf95ecf9747b47921f7ce7b865eb139f75

- The domain SHOULD be read from a trusted data source such as the browser window or over WalletConnect ([EIP-1328](https://eips.ethereum.org/EIPS/eip-1328)) sessions for comparison against the signing message contents.

#### Creating Sign-In with Ethereum interfaces
- Wallet implementers MUST display to the user the following terms from the Sign-In with Ethereum signing request by default and prior to signing, if they are present: `domain`, `address`, `statement`, and `resources`. Other present terms MUST also be made available to the user prior to signing either by default or through an extended interface.
- Wallet implementers displaying a plaintext `message` to the user SHOULD require the user to scroll to the bottom of the text area prior to signing.
- Wallet implementers MAY construct a custom Sign-In With Ethereum user interface by parsing the ABNF terms into data elements for use in the interface. The display rules above still apply to custom interfaces.

#### Supporting internationalization (i18n)
- After successfully parsing the message into ABNF terms, translation may happen at the UX level per human language.

## Rationale

### Requirements

> From the [RFP](https://notes.ethereum.org/@djrtwo/sign-in-with-ethereum-RFP):

Reviewer 9: The rationale section should explain the reasoning behind certain decisions in the EIP. Linking to the RFP is not appropriate here.

Author: Addressed in https://github.com/ethereum/EIPs/pull/4361/commits/e59daacf95ecf9747b47921f7ce7b865eb139f75

        - https://github.com/trezor/trezor-firmware/pull/1568
    - EIP-191 is simple to implement using a pre-set prefix prior to signing, while EIP-712 is more complex to implement requiring the further implementations of a bespoke Solidity-inspired type system, RLP-based encoding format, and custom keccak-based hashing scheme. **(2)**
    - EIP-191 produces more human-readable messages, while EIP-712 creates signing outputs for machine consumption, with most wallets not displaying the payload to be signed in a manner friendly to humans. **(1)**![](https://i.imgur.com/yvhz2mD.png)

    - EIP-712 has the advantage of on-chain representation and on-chain verifiability, such as for their use in metatransactions, but this feature is not relevant for the specification's scope. **(2)**
- Why not use JWTs? Wallets don't support JWTs. The keccak hash function is [not assigned by IANA](https://www.iana.org/assignments/jose/jose.xhtml) for use as a JOSE algorithm. **(2, 3)**
- Why not use YAML or YAML with exceptions? YAML is loose compared to ABNF, which can readily express character set limiting, required ordering, and strict whitespacing. **(2, 3)**

> ### Discussions and Acknowledgements

Reviewer 9: ```suggestion
### Discussions and Acknowledgements
```
This type of information should be tracked in the `discussions-to` thread, not in the EIP itself.

Author: Addressed in https://github.com/ethereum/EIPs/pull/4361/commits/e59daacf95ecf9747b47921f7ce7b865eb139f75

- @SamWilsn made several improvements to the ABNF grammar and additional security guidelines to wallet implementers to prevent phishing attacks.
- @cfelde suggested a minor improvement to the ABNF grammar.

### Out of Scope
The following concerns are out of scope for this version of the specification to define:
- Additional authentication not based on Ethereum addresses.
- Authorization to server resources.
- Interpretation of the URIs in the `resources` term as claims or other resources.
- The specific mechanisms to ensure domain-binding.
- The specific mechanisms to generate nonces and evaluation of their appropriateness.
- Protocols for use without TLS connections.

## Backwards Compatibility
- Most wallet implementations already support EIP-191, so this is used as a base pattern with additional features.
- Requirements were gathered from existing implementations of similar sign-in workflows, including statements to allow the user to accept a Terms of Service, nonces for replay protection, and inclusion of the Ethereum address itself in the message.

> ## Forwards Compatibility

Reviewer 9: ```suggestion
## Forwards Compatibility
```
This is not a valid EIP section. This sort of information should be incorporated into the rationale.

Reviewer 6: ![Screenshot 2021-12-06 at 16-19-50 Meme Generator - Imgflip](https://user-images.githubusercontent.com/57262657/144924359-bc467640-3814-4932-96bf-68ed0607f225.png)

#### Channel security
- For web-based applications, all communications SHOULD use HTTPS to prevent man-in-the-middle attacks on the message signing.
- When using protocols other than HTTPS, all communications SHOULD be protected with proper techniques to maintain confidentiality, data integrity, and sender/receiver authenticity.

#### Session invalidation
There are several cases where an implementer SHOULD check for state changes as they relate to sessions.

- If an EIP-1271 implementation or dependent data changes the signature computation, the server SHOULD invalidate sessions appropriately.
- If any resources specified in `resources` change, the server SHOULD invalidate sessions appropriately. However, the interpretation of `resources` is out of scope of this specification.

#### Maximum lengths for ABNF terms
- While this specification does not contain normative requirements around maximum string lengths, implementers SHOULD choose maximum lengths for terms that strike a balance across the prevention of denial of service attacks, support for arbitrary use cases, and user readability.

> ## Additional Resources

Reviewer 9: This is not a valid EIP section. Please remove.

Author: Addressed in https://github.com/ethereum/EIPs/pull/4361/commits/e59daacf95ecf9747b47921f7ce7b865eb139f75
